File: 704. Binary Search/binarysearch.java
class binarysearch {
    public int search(int[] nums, int target) {
        //this is the iterative approach. 
        int low = 0;
        int high = nums.length - 1;
        while(low <= high){
            int mid = (low + high) / 2;
            int cur = nums[mid];
            if(cur < target) low = mid + 1;
            else if(cur > target) high = mid - 1;
            else return mid;
        }
        return -1;
    }
}
//binarysearch s = new binarysearch();
//int[] nums = {-1,0,1,4,5};
//System.out.println(s.search(nums, 4));
//iterative approach

File: 704. Binary Search/binarysearch.py
class Solution(object):
    def binarySearch(self, nums, target, low, high):
        mid = (low + high) // 2
        cur = nums[mid]
        if low > high:
            return -1
        if cur < target:
            return self.binarySearch(nums, target, mid + 1, high)
        elif cur > target:
            return self.binarySearch(nums, target, low, mid - 1)
        elif cur == target:
            return mid

    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        low = 0
        high = len(nums) - 1
        return self.binarySearch(nums, target, low, high)


#s = Solution()
#nums = [-1, 0, 1, 5, 6]
#s.search(nums, 1)
# RECURSIVE APPROACH